feat: Read Only transactions via @frappe.read_only

gavindsouza opened the pull request:
When you wrap a function with `@frappe.read_only` and if you've done a replica setup, on calling said function a database connection with the replica server will be made in the context of the function's execution. When there's no replica setup, a separate connection will be started with the primary with read-only transactions.

Essentially, functions wrapped with `@frappe.read_only` will only allow reads in the context of the function. To call a function that allows writes from this, you'd have to wrap it with `@frappe.write_only`.


### Other Changes

* "perf" fixes for `frappe.db.check_transaction_status`
* Marked `frappe.render_template` as read only. This will disallow writes through it

gavindsouza committed:
refactor: Check Transaction Status

* Add ability to disallow writes when frappe.flags.read_only is set
* Restructured check for simplicity
* Minor perf improvements due to elimination of extra query reading and parsing

gavindsouza committed:
fix(minor): Set read_only when wrapped around @read_only

gavindsouza committed:
fix: Handle read only settings in @write_only

gavindsouza committed:
fix(utils): Cleaning and identifying queries

gavindsouza committed:
fix: Make frappe.render_template read only

* Writes shouldn't be allowed via frappe.render_template

gavindsouza committed:
feat: READ ONLY transactions

If replica setup not done, @read_only will default to starting a read
only transaction, in a new connection

Other changes:
* Removed need to scrub sql or identify if a read only query is executed
* Added frappe.db.transaction_started to keep track of if db.begin has
  to be run
* This runs a START TRANSACTION if not already run in each "transaction"

gavindsouza committed:
fix: Queries starting with WITH or SELECT as reads

gavindsouza committed:
fix: Start a new transaction if no writes are made so far

gavindsouza committed:
fix: Increase write count only if insert/update/delete

ankush commented:
> Marked frappe.render_template as read only. This will disallow writes through it


Read-only transactions are also good for overall performance, should enable this in other places too? 🚀 

> InnoDB can avoid the overhead associated with setting up the transaction ID (TRX_ID field) for transactions that are known to be read-only. A transaction ID is only needed for a transaction that might perform write operations or locking reads such as SELECT ... FOR UPDATE. Eliminating unnecessary transaction IDs reduces the size of internal data structures that are consulted each time a query or data change statement constructs a read view.

 Ref: https://dev.mysql.com/doc/refman/5.6/en/innodb-performance-ro-txn.html

gavindsouza commented:
From the CI issues, Access Log inserts seemed to be a part of the problem, decided to get it "fixed" by changing the table's engine to MyISAM, more performant for mindless quick inserts. No, haven't pushed that yet.

Apart from that, I might've hit a sort of a block with maintaining data consistency between the read-only and default session with respect to uncommitted data...still looking into ways to deal with this...Maybe I'll have to consider an alternate way to deal with this altogether.

----

> Read-only transactions are also good for overall performance, should enable this in other places too? 🚀

That's the idea 😄 , but only after we get this right 🥲

gavindsouza committed:
fix: Use connection if exists, close on destroy

gavindsouza committed:
perf: Change Access Log table engine to MyISAM

For this particular case, I think we want to optimize inserts.
Other Reads and updates are very rare if ever, this gets rid of all the lock and slow inserts we've been dealing with InnoDB now ;)

gavindsouza committed:
refactor: Re-arrange MariaDBDatabase into semantic blocks

gavindsouza committed:
refactor(Minor): Consistent DoesNotExistError messages

gavindsouza committed:
chore(test_website): Suppress unnecessary stdout of tests

gavindsouza committed:
fix: EXPLAIN is a read query

gavindsouza committed:
fix: Revert usage of transaction_started and using implicit begin

gavindsouza committed:
chore: Get rid of the redundant read_only flags

gavindsouza committed:
style: Remove extra indent

surajshetty3416 commented:
@gavindsouza Please re-open once this is ready for review.

gavindsouza commented:
This was a bit annoying/hard to do in terms of maintaining transaction consistency for single request - multiple DB sessions. Not sure if this is even a real possibility 😢
